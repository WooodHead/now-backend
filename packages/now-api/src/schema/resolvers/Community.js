import {
  defaultCursorSerializer,
  sqlPaginatify,
  userIdFromContext,
} from '../util';
import { Community, Membership } from '../../db/repos';
import { genRandomUuid, now } from '../../db/sql';
import { isAdmin } from '../AdminDirective';

const allCommunities = (root, { input, orderBy = 'id' }) =>
  sqlPaginatify(orderBy, Community.all({}), input);

const communityQuery = (root, { id }, context) =>
  (isAdmin(context)
    ? Community.byId(id)
    : Community.get({
        'communities.id': id,
        'memberships.userId': userIdFromContext(context),
      })
        .column('communities.*')
        .innerJoin('memberships', {
          'memberships.communityId': 'communities.id',
        })
  ).then(community => {
    context.loaders.communities.prime(id, community);
    return community;
  });

export const queries = { allCommunities, community: communityQuery };

const createCommunity = (root, { input: { name } }, { loaders }) =>
  Community.insert({ id: genRandomUuid(), name })
    .returning('id')
    .then(([id]) => ({
      community: loaders.communities.load(id),
    }));

const updateCommunity = async (
  root,
  { input: { id, ...fields } },
  { loaders }
) => {
  loaders.communities.clear(id);

  const updateCount = await Community.update({
    id,
    ...fields,
    updatedAt: now(),
  });

  if (updateCount === 0) {
    throw new Error('Community not found');
  }

  return {
    community: loaders.communities.load(id),
  };
};

export const mutations = { createCommunity, updateCommunity };

const users = ({ id: communityId }, { input = {} }, context) =>
  sqlPaginatify('id', Membership.all({ communityId }), {
    ...input,
    edgeBuilder: (cursorId, membership) => ({
      cursor: defaultCursorSerializer(cursorId, membership),
      node: context.loaders.members.load(membership.userId),
      ...membership,
    }),
  });

export const resolvers = { users };
