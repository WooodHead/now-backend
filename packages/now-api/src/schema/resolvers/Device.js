import { query, update, TABLES } from '../../db';
import { userQuery } from './User';
import { userIdFromContext } from '../util';
import { assocDevice } from '../../fcm';

export const getDevices = userId =>
  query({
    TableName: TABLES.DEVICE,
    KeyConditionExpression: 'userId = :userId',
    ExpressionAttributeValues: { ':userId': userId },
    IndexName: 'userId-index',
  });

const user = (device, args, context) =>
  userQuery(device, { id: device.userId }, context);

export const resolvers = { user };

const registerDevice = (root, { input: { token, type, model } }, context) => {
  const userId = userIdFromContext(context);
  const now = new Date().toISOString();

  return update(
    TABLES.DEVICE,
    { token },
    'set #t=:type, model=:model, userId=:userId, createdAt=if_not_exists(createdAt,:now), updatedAt=:now',
    {
      ':type': type,
      ':model': model,
      ':userId': userId,
      ':now': now,
    },
    { '#t': 'type' }
  ).then(r => {
    assocDevice(token, context.user);
    return { device: r.Attributes };
  });
};

export const mutations = { registerDevice };
