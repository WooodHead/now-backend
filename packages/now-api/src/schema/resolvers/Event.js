import { ChronoUnit, Instant, ZoneId } from 'js-joda';
import uuid from 'uuid/v4';
import { toNumber, isInteger } from 'lodash';

import { userIdFromContext, sqlPaginatify } from '../util';
import { getEventRsvps, userDidRsvp } from './Rsvp';
import { getMessages } from './Message';
import { getPubSub } from '../../subscriptions';
import { Event, EventUserMetadata, Rsvp } from '../../db/repos';
import sql from '../../db/sql';

const topicName = eventId => `event-changes-${eventId}`;

export const notifyEventChange = eventId =>
  getPubSub().publish(topicName(eventId), true);

// Resolvers
const activityResolver = ({ activityId }, args, { loaders }) =>
  loaders.activities.load(activityId);

const rsvpsResolver = (root, args) =>
  getEventRsvps({
    eventId: root.id,
    ...args,
  });

const messagesResolver = async (root, args) => {
  const eventId = root.id;

  const eventMessageConnection = await getMessages(root, {
    eventId,
    ...args,
  });

  return { eventId, ...eventMessageConnection };
};

const isAttendingResolver = ({ id }, { userId }, ctx) =>
  userDidRsvp({ eventId: id, userId: userId || userIdFromContext(ctx) });

// one day, this will be fancier.
const stateResolver = ({ time }) => {
  const eventTime = time.toInstant();
  const now = Instant.now();

  if (now.isBefore(eventTime)) return 'FUTURE';
  else if (eventTime.until(now, ChronoUnit.HOURS) < 2) return 'PRESENT';
  return 'PAST';
};

const locationResolver = ({ locationId }, args, { loaders }) =>
  loaders.locations.load(locationId);

const timeResolver = ({ time, timezone }) =>
  time
    .withZoneSameInstant(ZoneId.of(timezone))
    .truncatedTo(ChronoUnit.SECONDS)
    .withFixedOffsetZone();

export const resolvers = {
  activity: activityResolver,
  rsvps: rsvpsResolver,
  messages: messagesResolver,
  isAttending: isAttendingResolver,
  state: stateResolver,
  location: locationResolver,
  time: timeResolver,
};

// Queries
const allEvents = (root, { input, orderBy = 'id' }) =>
  sqlPaginatify(orderBy, Event.all({}), input);

const manyEvents = (root, { ids }, { loaders }) => loaders.events.loadMany(ids);

const eventQuery = (root, { id }, { loaders }) => loaders.events.load(id);

export const queries = { event: eventQuery, allEvents, manyEvents };

const createEvent = (
  root,
  { input: { time, timezone, activityId, limit, locationId } },
  { loaders }
) => {
  const newId = uuid();
  const newEvent = {
    id: newId,
    locationId,
    activityId,
    limit,
    time: time.toString(),
    timezone: timezone.id(),
    createdAt: sql.raw('now()'),
    updatedAt: sql.raw('now()'),
  };

  loaders.events.clear(newId);

  return Event.insert(newEvent).then(() => ({
    event: loaders.events.load(newId),
  }));
};

const updateEvent = (
  root,
  { input: { id, time, timezone, activityId, limit, locationId } },
  { loaders }
) => {
  const updatedEvent = {
    id,
    locationId,
    activityId,
    limit,
    time: time.toString(),
    timezone: timezone.id(),
    createdAt: sql.raw('now()'),
    updatedAt: sql.raw('now()'),
  };

  loaders.events.clear(id);

  return Event.update(updatedEvent).then(() => {
    notifyEventChange(id);
    return {
      event: loaders.events.load(id),
    };
  });
};

const markEventChatRead = async (root, { input: { eventId, ts } }, ctx) => {
  if (!isInteger(toNumber(ts))) {
    throw new Error('ts must be an integer as a string');
  }
  const userId = userIdFromContext(ctx);
  const potentialMetadata = await EventUserMetadata.get({ eventId, userId });
  let metadata = potentialMetadata;
  if (!metadata) {
    metadata = {
      id: uuid(),
      eventId,
      userId,
      createdAt: sql.raw('now()'),
    };
    await EventUserMetadata.insert(metadata);
  }

  const { id } = metadata;

  const updatedMetadata = {
    id,
    lastReadTs: ts,
    updatedAt: sql.raw('now()'),
  };

  ctx.loaders.events.clear(eventId);

  return EventUserMetadata.update(updatedMetadata).then(() => ({
    rsvp: () => Rsvp.get({ eventId, userId }),
    event: () => ctx.loaders.events.load(eventId),
  }));
};

export const mutations = {
  createEvent,
  updateEvent,
  markEventChatRead,
};

const eventSubscription = {
  subscribe: (root, { id }) => getPubSub().asyncIterator(topicName(id)),
  resolve: (payload, { id }, { loaders }) => loaders.events.load(id),
};

export const subscriptions = { event: eventSubscription };
