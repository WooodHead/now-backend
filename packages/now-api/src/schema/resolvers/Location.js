import { Location } from '../../db/repos';
import { sqlPaginatify } from '../util';
// Queries
const allLocations = (root, { input, orderBy = 'id' }) =>
  sqlPaginatify(orderBy, Location.all({}), input);

const manyLocations = (root, { ids }, { loaders }) =>
  loaders.locations.loadMany(ids);

const locationQuery = (root, { id }, { loaders }) => loaders.locations.load(id);

export const queries = { location: locationQuery, allLocations, manyLocations };

export const mutations = {};

export const resolvers = {};
