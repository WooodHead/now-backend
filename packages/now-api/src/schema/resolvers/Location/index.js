import uuid from 'uuid/v4';
import { Geometry, Point } from 'wkx';

import { now, prefixSearch } from '../../../db/sql';
import { Location } from '../../../db/repos';
import { sqlPaginatify } from '../../util';
import { syncWeworkMarket, weworkMarkets } from './wework';

// Queries
const allLocations = (root, { input, orderBy = 'id', prefix }) => {
  const args = prefix ? prefixSearch('name', prefix, true) : [];
  return sqlPaginatify(orderBy, Location.all(...args), input);
};

const manyLocations = (root, { ids }, { loaders }) =>
  loaders.locations.loadMany(ids);

const locationQuery = (root, { id }, { loaders }) => loaders.locations.load(id);

export const queries = {
  location: locationQuery,
  allLocations,
  manyLocations,
  weworkMarkets,
};

// visible for testing
export const parseInputGeometry = ({ lat, lng, geojson }, required = true) => {
  if (lat !== undefined || lng !== undefined) {
    if (geojson !== undefined) {
      throw new Error('Either lat/lng or geojson is required, but not both.');
    }
    if (typeof lat !== 'number' || Math.abs(lat) > 90) {
      throw new Error('Invalid latitude.');
    }
    if (typeof lng !== 'number' || Math.abs(lng) > 180) {
      throw new Error('Invalid longitude.');
    }

    return new Point(lng, lat).toWkt();
  }

  if (typeof geojson !== 'object') {
    if (typeof geojson === 'undefined' && !required) {
      return undefined;
    }
    throw new Error('Either lat/lng or geojson is required.');
  }

  let geom;
  try {
    geom = Geometry.parseGeoJSON(geojson);
  } catch (e) {
    throw new Error('Invalid GeoJSON');
  }
  if (!(geom instanceof Point)) {
    throw new Error('Only Point geometries are accepted.');
  }
  return geom.toWkt();
};

const createLocation = (root, { input }, { loaders }) => {
  const {
    foursquareVenueId,
    address,
    name,
    crossStreet,
    city,
    state,
    postalCode,
    country,
    neighborhood,
  } = input;

  const id = uuid();
  const location = parseInputGeometry(input);

  return Location.insert({
    id,
    foursquareVenueId: foursquareVenueId || null, // coerce empty string to null
    location,
    address,
    name,
    crossStreet,
    city,
    state,
    postalCode,
    country: country || 'United States',
    neighborhood,
    createdAt: now(),
    updatedAt: now(),
  }).then(() => ({ location: loaders.locations.load(id) }));
};

const updateLocation = (root, { input }, { loaders }) => {
  const {
    id,
    foursquareVenueId,
    address,
    name,
    crossStreet,
    city,
    state,
    postalCode,
    country,
    neighborhood,
  } = input;

  const location = parseInputGeometry(input, false);

  loaders.locations.clear(id);

  return Location.update({
    id,
    foursquareVenueId: foursquareVenueId || null, // coerce empty string to null
    location,
    address,
    name,
    crossStreet,
    city,
    state,
    postalCode,
    country: country || 'United States',
    neighborhood,
    updatedAt: now(),
  }).then(() => ({ location: loaders.locations.load(id) }));
};

export const mutations = { createLocation, updateLocation, syncWeworkMarket };

const lng = ({ location }) => location.x;
const lat = ({ location }) => location.y;
const geojson = ({ location }) => location.toGeoJSON();

export const resolvers = { lng, lat, geojson };
