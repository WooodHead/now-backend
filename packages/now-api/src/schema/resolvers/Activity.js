import uuid from 'uuid';
import { LocalDateTime, LocalTime, ZoneId, LocalDate } from 'js-joda';

import { sqlPaginatify } from '../util';
import { Activity, Event } from '../../db/repos';
import sql from '../../db/sql';
import { notifyEventChange } from './Event';

export const NYC_TZ = ZoneId.of('America/New_York');
export const AVAILABILITY_HOUR = LocalTime.parse('21:00');
export const EARLY_AVAILABILITY_HOUR = LocalTime.parse('20:00');

export const getToday = () => {
  const now = LocalDateTime.now(NYC_TZ);
  return (now.toLocalTime().isBefore(AVAILABILITY_HOUR)
    ? now.toLocalDate()
    : now.toLocalDate().plusDays(1)
  ).toString();
};

const allActivities = (root, { input, orderBy = 'id' }) =>
  sqlPaginatify(orderBy, Activity.all({}), input);

const manyActivities = (root, { ids }, { loaders }) =>
  loaders.activities.loadMany(ids);

const activityQuery = (root, { id }, { loaders }) =>
  loaders.activities.load(id);

const todayActivity = () => Activity.get({ activityDate: getToday() });

export const queries = {
  todayActivity,
  activity: activityQuery,
  allActivities,
  manyActivities,
};

export const getEvents = ({ id }, { first, last, after, before }) =>
  sqlPaginatify('id', Event.all({ activityId: id }), {
    first,
    last,
    after,
    before,
  });

export const resolvers = {
  events: getEvents,
  generallyAvailableAt: () =>
    LocalDate.now()
      .atTime(AVAILABILITY_HOUR)
      .atZone(NYC_TZ),
};

const createActivity = (
  root,
  { input: { title, description, activityDate, emoji, pushNotification } },
  { loaders }
) => {
  const newId = uuid.v1();
  const newActivity = {
    id: newId,
    title,
    pushNotification,
    description,
    activityDate: activityDate.toString(),
    createdAt: sql.raw('now()'),
    updatedAt: sql.raw('now()'),
    emoji,
  };

  loaders.activities.clear(newId);

  return Activity.insert(newActivity).then(() => ({
    activity: loaders.activities.load(newId),
  }));
};

const updateActivity = async (
  root,
  { input: { id, title, description, activityDate, emoji, pushNotification } },
  { loaders }
) => {
  const updatedActivity = {
    id,
    title,
    pushNotification,
    description,
    activityDate: activityDate.toString(),
    updatedAt: sql.raw('now()'),
    emoji,
  };

  loaders.activities.clear(id);

  await Activity.update(updatedActivity);
  const activityEvents = await Event.all({ activityId: id });

  activityEvents.forEach(({ id: eventId }) => {
    notifyEventChange(eventId);
  });

  return {
    activity: loaders.activities.load(id),
  };
};

export const mutations = { createActivity, updateActivity };
