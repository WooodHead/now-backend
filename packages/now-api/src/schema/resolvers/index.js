import { GraphQLDate, GraphQLTime, GraphQLDateTime } from 'graphql-iso-date';
import { GraphQLUpload } from 'apollo-upload-server';
import GraphQLJSON from 'graphql-type-json';
import * as Device from './Device';
import * as Event from './Event';
import * as User from './User';
import * as Activity from './Activity';
import * as Message from './Message';
import * as Rsvp from './Rsvp';
import * as Photo from './Photo';
import * as Report from './Report';
import Name from './Name';
import Birthdate from './Birthdate';
import { wrapResolvers } from '../authorization';

export default wrapResolvers({
  Query: {
    ...User.queries,
    ...Activity.queries,
    ...Event.queries,
    ...Message.queries,
  },
  Subscription: {
    ...Event.subscriptions,
    ...Message.subscriptions,
  },
  Mutation: {
    ...Activity.mutations,
    ...Device.mutations,
    ...Event.mutations,
    ...Message.mutations,
    ...Rsvp.mutations,
    ...Photo.mutations,
    ...User.mutations,
    ...Report.mutations,
  },
  Activity: Activity.resolvers,
  Device: Device.resolvers,
  Event: Event.resolvers,
  Message: Message.resolvers,
  Rsvp: Rsvp.resolvers,
  User: User.resolvers,
  EventMessagesEdge: {
    node: root => ({
      ...root.node,
    }),
  },
  GraphQLDate,
  GraphQLTime,
  GraphQLDateTime,
  GraphQLJSON,
  Upload: GraphQLUpload,
  Name,
  Birthdate,
});
