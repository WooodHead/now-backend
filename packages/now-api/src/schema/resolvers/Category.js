import { tempTemplates } from './Template';
import { buildEdge } from '../../schema/util';

const tempCategories = {
  'd2ec5baa-638e-44f1-88d3-a67af0e1e9d0': {
    id: 'd2ec5baa-638e-44f1-88d3-a67af0e1e9d0',
    title: 'Explore',
    description:
      'Choose the activity that best allows you to discover new places, events, food, and more with others.',
    templateIds: [
      'ebf997e5-cf24-407f-b98d-5108d73b4044',
      'f2a7b3a5-b00f-4b23-a224-2db4049fec27',
    ],
  },
  'ec9c43f0-bdaf-4d5d-bfdc-830cc38ad50d': {
    id: 'ec9c43f0-bdaf-4d5d-bfdc-830cc38ad50d',
    title: 'Watch',
    description:
      'Choose the activity that best allows you to connect with others through shows, movies, concerts, and more.',
    templateIds: [
      'ef7ec7d3-6a3b-4c63-8450-ac5c1c3ce819',
      'ff18f172-74c3-4bcc-9c3e-45966c721869',
    ],
  },
  '5e27e0dc-5491-4dcf-99de-e33e20c13eeb': {
    id: '5e27e0dc-5491-4dcf-99de-e33e20c13eeb',
    title: 'Discuss',
    description:
      'Choose the activity that best allows you to connect with others through conversation.',
    templateIds: [
      '8f0f400f-0cd5-43dd-b357-3cc3f07be2bf',
      '0b721d9e-ead6-4a35-9b98-82b780ebd32a',
      'c26283a6-8aac-46ba-862c-889c42817dfa',
      'b5096d9f-7bba-4259-9bd2-2b2497e361b4',
    ],
  },
  'c5d89a42-3b04-43ea-90ae-e7a4e27aff5d': {
    id: 'c5d89a42-3b04-43ea-90ae-e7a4e27aff5d',
    title: 'Play',
    description:
      'Choose the activity that best allows you to connect with others through fun competition.',
    templateIds: [
      '6deb37cb-9f39-4ede-936a-20339a1b258c',
      '54d49999-4731-48ee-88a1-db8fe803f048',
    ],
  },
  '20cb2104-3f4a-477e-b792-e5595b9c5069': {
    id: '20cb2104-3f4a-477e-b792-e5595b9c5069',
    title: 'Learn',
    description:
      'Choose the activity that allows you to connect with others through sharing your skills, hobbies, and more.',
    templateIds: [
      'f6676ffa-f182-4067-9138-35223a6bf4e8',
      '46b8e494-bdc9-4b71-8e11-4cb325e01df9',
    ],
  },
  '716157ed-64a8-4036-ae88-a4c6cf19b04a': {
    id: '716157ed-64a8-4036-ae88-a4c6cf19b04a',
    title: 'Hang Out',
    description:
      'Choose the activity that best allows you to connect with others through social gatherings.',
    templateIds: ['d95e9bc7-dd14-48de-a800-e7fc1fe28228'],
  },
};

export const categoryQuery = id => tempCategories[id];

const getCategory = (root, { id }) => categoryQuery(id);

const getCategories = () => Object.values(tempCategories);

const getAllCategories = (root, { orderBy = 'title' }) => {
  const data = Object.values(tempCategories).sort(
    (a, b) => a[orderBy] > b[orderBy]
  );
  return {
    pageInfo: {
      hasNextPage: false,
      hasPreviousPage: false,
    },
    count: data.length,
    edges: data.map(d => buildEdge(orderBy, d)),
  };
};

export const queries = {
  category: getCategory,
  categories: getCategories,
  allCategories: getAllCategories,
};

const templates = ({ templateIds }) => templateIds.map(id => tempTemplates[id]);

export const resolvers = { templates };
