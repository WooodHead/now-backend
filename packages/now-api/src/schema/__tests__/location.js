import { parseInputGeometry } from '../resolvers/Location';

describe('parseInputGeometry', () => {
  it('rejects invalid inputs', () => {
    expect(() =>
      parseInputGeometry({ lat: 45 })
    ).toThrowErrorMatchingSnapshot();
    expect(() =>
      parseInputGeometry({ lat: 100, lng: 45 })
    ).toThrowErrorMatchingSnapshot();
    expect(() =>
      parseInputGeometry({ lat: 45, lng: 20, geojson: {} })
    ).toThrowErrorMatchingSnapshot();
    expect(() =>
      parseInputGeometry({
        geojson: {
          type: 'LineString',
          coordinates: [[-170, 10], [170, 11]],
        },
      })
    ).toThrowErrorMatchingSnapshot();
  });

  it('parses lat/lng', () => {
    expect(
      parseInputGeometry({ lng: -73.9961013793945, lat: 40.7256011962891 })
    ).toMatchSnapshot();
  });

  it('parses geojson', () => {
    expect(
      parseInputGeometry({
        geojson: {
          type: 'Point',
          coordinates: [-73.9961013793945, 40.7256011962891],
        },
      })
    ).toMatchSnapshot();
  });

  it('handles empty input when told to', () => {
    expect(() => parseInputGeometry({})).toThrowErrorMatchingSnapshot();
    expect(parseInputGeometry({}, false)).toBeUndefined();
  });
});
