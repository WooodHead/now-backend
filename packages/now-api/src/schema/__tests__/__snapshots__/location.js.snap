// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseInputGeometry handles empty input when told to 1`] = `"Either lat/lng or geojson is required."`;

exports[`parseInputGeometry parses geojson 1`] = `"POINT(-73.9961013793945 40.7256011962891)"`;

exports[`parseInputGeometry parses lat/lng 1`] = `"POINT(-73.9961013793945 40.7256011962891)"`;

exports[`parseInputGeometry rejects invalid inputs 1`] = `"Invalid longitude."`;

exports[`parseInputGeometry rejects invalid inputs 2`] = `"Invalid latitude."`;

exports[`parseInputGeometry rejects invalid inputs 3`] = `"Either lat/lng or geojson is required, but not both."`;

exports[`parseInputGeometry rejects invalid inputs 4`] = `"Only Point geometries are accepted."`;
